crates/bevy_pbr/src/light.rs

```diff
@@ -52,6 +53,84 @@ impl PointLightUniform {
     }
 }
 
+/// A Directional light.
+///
+/// Directional lights don't exist in reality but they are a good
+/// approximation for light sources VERY far away, like the sun or
+/// the moon.
+#[derive(Debug, Reflect)]
+#[reflect(Component)]
+pub struct DirectionalLight {
+    pub color: Color,
+    pub intensity: f32,
+    direction: Vec3,
+}
+
+impl DirectionalLight {
+    /// Create a new directional light component.
+    ///
+    /// # Panics
+    /// Will panic if `direction` is not normalized.
+    pub fn new(color: Color, intensity: f32, direction: Vec3) -> Self {
+        assert!((direction.length_squared() - 1.0).abs() < 0.0001);
+        DirectionalLight {
+            color,
+            intensity,
+            direction,
+        }
+    }
+
+    /// Set direction of light.
+    ///
+    /// # Panics
+    /// Will panic if `direction` is not normalized.
+    pub fn set_direction(&mut self, direction: Vec3) {
+        assert!((direction.length_squared() - 1.0).abs() < 0.0001);
```

Comment:
The cost of checking for normalization is almost as expensive as just normalizing. That also removes a source of "random" runtime panics, which should be avoided if possible.

Reply:
"almost" being of course defined by your perspective. Of course sqrt + 3 more divides is "more expensive".

Reply:
Alternatively we could just print a warning.

Reply:
Or we could normalize in the shader. I don't see much harm in letting the gpu just do its thing.

Reply:
What are the thoughts on having a strongly typed `NormalizedVec3`? 
This forces the caller to ensure its normalized and we don't have worry checking is it's normalized/normalizing the vector.

Reply:
I think it would be great if it could be upstreamed in Glam

Reply:
Agreed, however in the meantime is it worth implementing a `Normalized<Vec3>` and then if/when it gets upstreamed into Glam, use the Glam implementation?

Reply:
I see a great deal of possible misuse with a Normalized<Vec3> type, as it will probably lead to unnecessary normalization all over the place...

Reply:
Normalizing on the GPU would be a very bad idea. It is indeed very fast at doing so, but it's still much lower overall to do the computation hundreds of thousands of times instead of just once. (it is done per pixel after all)

Reply:
> I see a great deal of possible misuse with a Normalized<Vec3> type

Um, I don't particularly see how this can be misused? The entire purpose of the type is to guarantee that the vector is normalized so that if you take a `Normalized<Vec3>` you don't have to worry about checking if it is normalized since all of that is left to the caller of the function.
This is akin to `std::ptr::NonNull`.

Reply:
Granted, this is probably out of scope for the PR, so I'm still happy with the current implementation :smile:
(I just might make a PR later to add a guaranteed-normalized type) :)

Reply:
I'm not pointing at the mathematical correctness of the thing, I'm pointing at the performance. Unless you know of a better way, to ensure the value is always normalized just means that any change to it will always be normalized, even if it already is. 

So the correct (not slow) way of using such a type is to put it only at very specific points where the input value might not be and then ensure it's kept untouched all the way through the pipeline. It is, in my opinion, much more correct (fast and mathematically correct) to only put normalization where it's actually needed.

Reply:
There is something very unclear to me here.

In [my previous PR for smooth_damp](https://github.com/bevyengine/bevy/pull/2001), @NathanSWard pointed out how f64::clamp panics if min > max and that panicing was the rust way, so I ended up complying and changed my max on the inputs to asserts. 

Here, @cart, you seem more inclined to not go the rust way and have the input made safe on behalf of the user with a normalize call which is much more expansive than a simple max. I'm also more inclined to do so as, to me, panic should be only used on much more worrying stuff.

The only reason I've put an assert here is that. I'm super inclined into removing it and just normalizing the input vector, but then I'll also remove the asserts in smooth_damp and just max(0,input)

Reply:
By the `rust way` is was mostly referring to how the rust standard library implements a lot of it's functionality. That doesn't means we have to follow this, but rather it was a suggestion :smile:

Currently this is how I see the optional paths forward. (Regardless of which path is taken it should be documented of course)
1) `assert!`:
2) `normalize` the vector: this is nice because there are no runtime panics, but we have to normalize the vector even if it is alread, therefore doing extra work. 
3) do neither and say that the behavior is `undefined` if the vector is not normalized. (How this would require marking the function `unsafe` and adding a `debug_assert`. 

Of the 3 options I would prefer 1 or 2. 

_note_: this is why I'm advocating for the `NormalizedVec3` as the input parameter since this means we don't have to panic, normalize, or unsafe the function.

Reply:
Ok :)
I'll go the easy, safe and pretty fast way : normalize the input. Same goes for smooth_damp, I'll just max(0,input)